_posts/2024-03-13-bitcoin-inquisition-#45.md

```diff
@@ -11,17 +11,89 @@ status: upcoming
+  unchanged.
+
+The semantics of the signature-checking opcodes are similar to Tapscript
+opcodes, including the empty signature->false behavior for
+`OP_CHECKSIGFROMSTACK`.
+
+### [The code](https://github.com/bitcoin-core-review-club/bitcoin/tree/prbitcoin-inquisition-45)
+
+#### `OP_INTERNALKEY`
+
+Because this PR is for bitcoin inquisition, the
+[implementation](https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/script/interpreter.cpp#L1309)
+of `OP_INTERNALKEY` is remarkably simple. The taproot internal key is already
+extracted for use with verifying tapscript `OP_CHECKSIG(VERIFY|ADD)` with the
+key `0x01`, so it simply needs to be added to the stack.
+
+#### `OP_CHECKSIGFROMSTACK(VERIFY)`
+
+The implementation of CSFS(V) is split between the [main opcode
+processing](https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/script/interpreter.cpp#L1318), and a function analogous to `EvalChecksigPreTapscript` and `EvalChecksigTapscript` named [`EvalChecksigFromStack`](https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/script/interpreter.cpp#L346).
 
+My original approach was to add a `SignatureChecker` that could carry the
+message rather than generating the message from the transaction, and give CSFS
+the exact semantics of existing sigops in their respective script types, but
+that approach had two problems: getting BIP340 signing into legacy/segwitv0
+scripts became intrusive, and the semantics of legacy/segwitv0 signature
+checking are not particularly desirable for a new opcode. As a result I
+changed to this implementation that uses a separate function specific to CSFS
+and gives modern Tapscript-style semantics. Only after changing code
+approaches did I realize that ECDSA checking might not be desirable in these
+new sigops. The commit history still reflects the removal of ECDSA in a
+separate commit.
+
+Outside of these changes in `interpreter.cpp`, the only non-flag change in
+production code is to allow variable length input to BIP340
+signature-checking.
+
+### Testing
+
+For both INTERNALKEY and CSFS I wrote tests in the Transaction Test framework:
+
+https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/test/data/tx_valid.json#L671
+https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/test/data/tx_invalid.json#L491
+
+You can run a regtest node from this branch to use them yourself.
+
+I posted some scripts using `bitcoinjs-lib` and `noble-curves` that I used to
+generate the test vectors above:
+https://gist.github.com/reardencode/9dbc60a6d6e1591905d25bf4d123dfdd
 
 ## Questions
 
 1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?
--->
+
+## Open considerations
```

Comment:
I'd recommend adding some questions to help people check their understanding. Softballs like "which nop is being repurposed" / "what does this opcode do" or conceptual questions "what's the motivation for this change" / "how is this useful" are good places to start. Otherwise the meeting will be very short...